### DISCUSSION TITLE
diesel migration 'collapse' command

### DISCUSSION THREAD

**Discussion by thomasmost** [diesel-rs/diesel#3298] [Ideas]
# Diesel Migration Collapse
The big idea is: add a CLI command for replacing your migrations directory with a single schema initialization migration that implements your full database schema as defined by your last migration.

## Why?
Just as we shouldn't leave commented out code lying around our repository forever, migrations should be preserved in version control and periodically collapsed. 

## Implementation
1. Create a dump file of the database, not including data
2. Delete all migrations and replace with a new one called `initialize` that implements the 'schema dump' from step 1
3. Drop records from the migration table and replace with a single record confirming the database initialized

### Before
```
2022-08-10-233328_add_company_table
|- up.sql
|- down.sql
2022-08-19-233328_add_user_table
|- up.sql
|- down.sql
```

### Command
`diesel migration collapse`

### After
```
2022-08-30-233328_initialize
|- up.sql
|- down.sql
```

## Feedback Requested
* Is this a bad idea or a poor practice?
* Is this more easily done manually, i.e. should not be added as a core feature to the CLI?
* Any implementation notes
* Any additional questions or comments

**Comment by thomasmost** [comment]
@weiznich or @sgrif — Curious what you think of this idea?

**Comment by weiznich** [comment]
I feel this should not be part of diesel cli as it is at least for me unclear how this would interact with anything other than the file based migrations. For example it seems to be dangerous to use something like that with an existing installation that uses `embed_migrations!`.

**Comment by thomasmost** [comment]
Got it. Make sense, we can consider this closed then.
